src/wallet/mod.rs

```diff
@@ -162,21 +162,108 @@ where
+    /// value earlier than the [`crate::blockchain::Blockchain`] stop_gap (default is 20) then a
+    /// larger stop_gap should be used to monitor for all possibly used addresses.  
+    Reset(u32),
+}
+
 // offline actions, always available
 impl<B, D> Wallet<B, D>
 where
     D: BatchDatabase,
 {
-    /// Return a newly generated address using the external descriptor
-    pub fn get_new_address(&self) -> Result<Address, Error> {
-        let index = self.fetch_and_increment_index(KeychainKind::External)?;
+    // Return a newly derived address using the external descriptor
+    fn get_new_address(&self) -> Result<Address, Error> {
+        let incremented_index = self.fetch_and_increment_index(KeychainKind::External)?;
+
+        self.descriptor
+            .as_derived(incremented_index, &self.secp)
+            .address(self.network)
+            .map_err(|_| Error::ScriptDoesntHaveAddressForm)
+    }
+
+    // Return the the last previously derived address if it has not been used in a received
+    // transaction. Otherwise return a new address using [`Wallet::get_new_address`].
+    fn get_unused_address(&self) -> Result<Address, Error> {
+        let current_index = self.fetch_index(KeychainKind::External)?;
+
+        let derived_key = self.descriptor.as_derived(current_index, &self.secp);
+
+        let script_pubkey = derived_key.script_pubkey();
+
+        let found_used = self
+            .list_transactions(true)?
+            .iter()
+            .flat_map(|tx_details| tx_details.transaction.as_ref())
+            .flat_map(|tx| tx.output.iter())
+            .any(|o| o.script_pubkey == script_pubkey);
+
+        if found_used {
+            self.get_new_address()
+        } else {
+            derived_key
+                .address(self.network)
+                .map_err(|_| Error::ScriptDoesntHaveAddressForm)
+        }
+    }
+
+    // Return derived address for the external descriptor at a specific index
+    fn peek_address(&self, index: u32) -> Result<Address, Error> {
+        self.descriptor
+            .as_derived(index, &self.secp)
+            .address(self.network)
+            .map_err(|_| Error::ScriptDoesntHaveAddressForm)
+    }
+
+    // Return derived address for the external descriptor at a specific index and reset current
+    // address index
+    fn reset_address(&self, index: u32) -> Result<Address, Error> {
+        self.set_index(KeychainKind::External, index)?;
 
         self.descriptor
             .as_derived(index, &self.secp)
             .address(self.network)
             .map_err(|_| Error::ScriptDoesntHaveAddressForm)
     }
 
+    /// Return a derived address using the external descriptor, see [`AddressIndex`] for
+    /// available address index selection strategies. If the wallet descriptor is not derivable
+    /// (ie. does not end with /*) then the same address will always be returned for any [`AddressIndex`].
```

Comment:
Maybe here it would be more appropriate to say:

> If none of the keys in the descriptor are derivable  ...

because it's not the descriptor itself that ends with `/*`

Reply:
Sounds good, fixed.